Gutenberg: Tiled gallery block

simison opened the pull request:
Rewrite of the Tiled gallery block.

Apologies for a huge PR. 🙇‍♂️ This is quite a beast of a block! Complete rewrites are a bit tricky to do in portions. Probably good if you review files without the diff.

_(I updated the description here so comments below might seem a bit out of context)_

# Screenshots

<img width="1083" alt="screen shot 2018-10-23 at 16 03 27" src="https://user-images.githubusercontent.com/87168/47362598-69721200-d6dd-11e8-8c65-06e1322c0871.png">
<img width="557" alt="screen shot 2018-10-23 at 16 02 16" src="https://user-images.githubusercontent.com/87168/47362600-69721200-d6dd-11e8-99f6-f47648c5729b.png">
<img width="912" alt="screen shot 2018-10-23 at 16 01 44" src="https://user-images.githubusercontent.com/87168/47362601-69721200-d6dd-11e8-8dd5-a80b7ecaca03.png">

# Testing & features

* [Spin up Jurassic Ninja testing site](https://jurassic.ninja/create?shortlived&jetpack-beta&branch=update/gutenpack-deps-tiled-gallery&gutenpack&gutenberg&calypsobranch=update/tiled-gallery-updates)
* Alternatively if you use local Jetpack setup, you can build the Jetpack preset:
    ```
    npm run sdk gutenberg client/gutenberg/extensions/presets/jetpack -- --output-dir=/path/to/jetpack/_inc/blocks
    ```
    — Just remember to `npm install` first since this adds new deps.
* Connect Jetpack
* Create a new post
* Insert Tiled Gallery block

### Confirm that these work as expected:

* You can resize the browser window, toggle Gutenberg editor sidebar, change the mobile browser orientation and the gallery will adapt to new size both at the theme and in the editor
* You can switch styles from the toolbar
    * Mosaic and columns layouts are missing; they just render squares -layout
* You can write captions for each image: circle layout doesn't support captions, though.
* You can remove images from small "x" or pressing backspace while the image is selected
* Different block-alignments and block-widths; make sure to test with a theme that supports them, such as [Suki](https://wordpress.org/themes/suki/) or [Twentynineteen](https://github.com/WordPress/twentynineteen)
* You can add and remove images using buttons as well via drag'n'drop
* You can save the post, refresh the page, and tiled gallery block still loads
* Saved tiled galleries with different layouts, columns and other settings as well with captions render at the theme side without issues
* You can change a number of columns from the sidebar and see the layout change: only squares and circles support "columns"; for other layouts, columns-setting is disabled.
* You can transform tiled gallery to/from core-gallery block and to image blocks.
* Enable lazy loading images & Photon CDN from Jetpack settings and confirm that the gallery plays nicely with those features

### Known issues
* Left+right block-alignments have some silly bug 
* Previews in style picker look wonky but probably acceptable for v1
* Block does not transform from `[gallery]` block-code and I'm not sure why. Might be colliding with core-gallery since these share the same shortcode.


# Technical specs

Requires Jetpack side PR: https://github.com/Automattic/jetpack/pull/10375

- I copied over [Core gallery](https://github.com/WordPress/gutenberg/tree/5f01719c2319310132fd13084056d9aa75bfd2c8/packages/block-library/src/gallery) from Gutenberg as a starting point
   - There are some areas of code that tiled gallery could share with core-gallery, but I haven't done much about it yet. I've kept `gallery-image.js` exactly as it's Gutenberg so that we could use it as a component from Gutenberg later on. Also `getActiveStyle()` helper is directly from Gutenberg.
  - From the previous trial version of tiled-gallery, I basically kept just the math for square tiles layout, simplifying the object that the math-function returns.
- Square tiles don't really need "rows" HTML nodes, but I've kept them here for simplicity since mosaic and columns layouts depend on them.
- Uses `ResizeObserver` to fire events when the gallery's wrapper element has size changes. Changes could be caused by mobile screen orientation changes, changes in the browser window, by appearing elements such as Gutenberg sidebar etc.
   - [Only Chrome supports it](https://www.caniuse.com/#feat=resizeobserver), so I'm using [a polyfill](https://github.com/que-etc/resize-observer-polyfill).
    - `ResizeObserver` API is really simple and it's performance is great because it doesn't need to be polling for size changes or fire on every `window.onResize`. [Read more](https://developers.google.com/web/updates/2016/10/resizeobserver) from Google and or check out [the spec](https://wicg.github.io/ResizeObserver/).
    - We can ignore all the usual viewport events these implementations typically listen to, as well all the events changing the editor size in Gutenberg. 
    - Resize events are throttled by 200ms both at the theme and editor, no matter how many editor blocks we need to redraw.
- Theme side `view.js` _does not rely on React._ I purposefully choose to implement it as vanilla JS to keep payload smaller, albeit causing a bit more maintenance for us. I explored [rendering with React](https://gist.github.com/simison/71dc8ca81d08616e56d3187198a9bcb7) as well so if we decide so, it would be quick enough change to that implementation instead.

## Missing two layouts

So Mosaic and Columns is missing. Squares and circles are in. 

We could decide to:

- Port math over [from PHP](https://github.com/Automattic/jetpack/tree/e07addf220d02226f04d874b1b565ceb9f502839/modules/tiled-gallery/tiled-gallery) — I've been looking into this but quite badly blocked by it.
- Use external JS library to calculate same or similar layouts.
- Use Gutenberg's [ServerSideRender](https://github.com/WordPress/gutenberg/tree/4b23d085f9e4f0a43d53ce66093ed24fdf5edc87/packages/components/src/server-side-render) component to render the two missing layouts via shortcode, as a temporary solution.

mtias commented:
I'd definitely go for style picker for the variations — that's what it's meant for. It ensures it's accessible even with sidebar collapsed.

MichaelArestad commented:
> Random order - The tiled gallery has "random" image order option, which will shuffle images to different order on each page refresh. Should we include that feature here?

It depends. If the order of images allows the user to rearrange them, then we should probably not randomize them. However, if in any of the layouts, we can't allow dragging to rearrange, then random order would be good.

> Captions - Should we include the animation in the block? I'd vote to drop that and do what ever core gallery does.

I agree with you. Let's drop those in favor of the way core gallery handles captions.

> Style picker 

Let's definitely go with the style picker for variations. I think they're better at conveying meaning than a 20x20 icon and, as Matias mentioned, it's what they're designed for.

> Circle layout issues

I think we should likely disable captions on the circles altogether. I'm not sure we want to mess with the blue border, though. @mtias what do you think?

> Margin between the tiles is 2px. Should user be able to adjust that?

How easy is that to do? Is the mosaic script fast enough for that to work well?

> Icon

[This](https://cloudup.com/ix68wrI5UB3
) was from the earlier rounds of design, but might work.

mtias commented:
> I think we should likely disable captions on the circles altogether.

Seems sensible to me.

mtias commented:
By the way, there's an idea about _duplicating_ block styles in a panel on the sidebar that would automatically make these available in both places, which I think would be great.

MichaelArestad commented:
I tried testing this today, but couldn't get it to show up in Jurassic Ninja (I did connect Jetpack). Ideas?

simison commented:
@MichaelArestad 
>I tried testing this today, but couldn't get it to show up in Jurassic Ninja (I did connect Jetpack). Ideas?

Looks like the JN link was a bit outdated (this now requires a branch from Jetpack, too). I updated and tested it works now. 👍 Thanks for the heads up.

I fixed a couple smaller bugs with circle -layout, rebased+squashed commits so this would be good for code review now. This is not _ready_ block but I'd like to leave following work to different PRs to make reviewing them easier & more focused.

simison commented:
I found some issues with `wp_kses` and `<style>` so I need to make some refactoring before this is good for review again.

simison commented on client/gutenberg/extensions/tiled-gallery/layouts.js:
```diff
@@ -0,0 +1,131 @@
+
+// @TODO
+const rectangularLayout = options => squareLayout( options );
+const columnsLayout = options => squareLayout( options );
```
Note how columns and rectangular (="mosaic") simply implement square layouts for now.

simison commented on client/gutenberg/extensions/tiled-gallery/constants.js:
```diff
@@ -0,0 +1,34 @@
+		label: _x( 'Circles', 'Tiled gallery layout', 'jetpack' ),
+		name: 'circle',
+	},
+];
```
These Gutenberg style definitions are used by [`getActiveStyle`](https://github.com/Automattic/wp-calypso/pull/27458/files#diff-aba368b720f7450e330970e2f15499daR103) also at `view.js`, thus making the frontend dependent on `wp.i18n` althought we're not rendering any strings there. That's probably okay for 1st iteration but need to look into this soon.

simison commented:
Rewrote, rebased, updated the description etc. 

Needs a design+code review as well as plenty of testing with different devices! 💪

mapk commented:
After testing, I have these notes...

**1. Uploaded images not showing**
After uploading images through the gallery block, I see them in the block, but I don't see them when I click to `edit` the block. Not sure what the disconnect is here, but I was confused.

![images-gone](https://user-images.githubusercontent.com/617986/47443635-a71c7c80-d769-11e8-8f15-184105be8cbd.gif)

----

**2. Tiled Gallery icon**
I think we should use a different icon for the Tiled Gallery. Right now it looks like this:
![screen shot 2018-10-24 at 8 49 56 am](https://user-images.githubusercontent.com/617986/47443908-2d38c300-d76a-11e8-8abd-acac6c0d63af.png)

Maybe we can use this one instead?
https://material.io/tools/icons/?icon=dashboard&style=outline
![screen shot 2018-10-24 at 8 53 21 am](https://user-images.githubusercontent.com/617986/47443965-4772a100-d76a-11e8-931d-11f50643251a.png)

----

**3. Difference between Tiled Columns, Tiled Mosaic, and Square tiles?**
I don't see any difference between these options in the style dropdown. Is it just b/c of the number of images I have at the moment?

![tile-styles](https://user-images.githubusercontent.com/617986/47444596-b0a6e400-d76b-11e8-8212-e4d7b2c95ca9.gif)

----

**4. Mobile worked really well**

simison commented:
This block should be refactored to use `styles`  block type option added in Gutenberg 4.2: https://github.com/WordPress/gutenberg/blob/5dd85b307a95a5f32e0d31f0fc60cf73aba7ee3d/docs/block-api.md#styles-optional

simison commented:
Important change to bring over for accessibility: https://github.com/WordPress/gutenberg/pull/11205

melchoyce commented:
Just tested. My notes:

1. I typed "Gallery" and then clicked on the Gallery block just as my brain processed that "Tiled Gallery" was a separate block — I was expecting them to be the same block, and for Tiled to be additional style options within the Gallery block itself. I can probably train myself out of this mindset in the future.
2. I'm glad I was able to transform from a regular to a Tiled gallery — I immediately went to do that next and was happy to see it as an option.
3. The caption overlay shows up even when I'm not focused inside of an image. Like Gallery, it should only appear when I have an image selected, and it should only appear on that particular image. The font size is also larger than the core Gallery block: 16 vs 13px. We should be consistent. 
4. I can't adjust the number of columns and I'm not sure why? The text field is disabled, and trying to drag the slider doesn't work. 
5. I can't seem to change between Mosaic, Columns, and Tiles — they all display the same layout.
6. I managed to do this by toggling off "crop images": 
![image](https://user-images.githubusercontent.com/2846578/48573515-692ef600-e8da-11e8-8896-4a7fc097d817.png)

I also agree with @mapk re:
> Maybe we can use this one instead?
> https://material.io/tools/icons/?icon=dashboard&style=outline

simison commented:
Iteration on upload-button: https://github.com/WordPress/gutenberg/pull/9682

sirreal commented on client/gutenberg/extensions/tiled-gallery/constants.js:
```diff
@@ -0,0 +1,34 @@
+		name: 'circle',
+	},
+];
+export const LAYOUTS = [ 'rectangular', 'columns', 'square', 'circle' ];
```
Is this duplicating the above array in large part? I'd like to avoid that if possible.

sirreal commented on client/gutenberg/extensions/tiled-gallery/edit.jsx:
```diff
@@ -3,58 +3,72 @@
+	InspectorControls,
+	MediaPlaceholder,
+	mediaUpload,
+	MediaUpload,
```
`mediaUpload` and `MediaUpload` 😵

simison replied:
Gutenberg `:shruggutenbergdealwithit:`

sirreal commented on client/gutenberg/extensions/tiled-gallery/edit.jsx:
```diff
@@ -3,58 +3,72 @@
+	}
+
+	return {
+		...pick( image, [ 'alt', 'id', 'link', 'url' ] ),
```
Creating a couple of extra objects here, we can improve with `Object.assign`

sirreal commented on client/gutenberg/extensions/tiled-gallery/edit.jsx:
```diff
@@ -3,58 +3,72 @@
+		this.onRemoveImage = this.onRemoveImage.bind( this );
 		this.setImageAttributes = this.setImageAttributes.bind( this );
 		this.addFiles = this.addFiles.bind( this );
+		this.uploadFromFiles = this.uploadFromFiles.bind( this );
```
Let's use arrows instead of bind, calypso style.

sirreal commented on client/gutenberg/extensions/tiled-gallery/edit.jsx:
```diff
@@ -3,58 +3,72 @@
 		this.state = {
 			selectedImage: null,
+			layout: getActiveStyleName( arguments[ 0 ].className ),
 		};
```
Instance property

sirreal commented on client/gutenberg/extensions/tiled-gallery/edit.jsx:
```diff
@@ -3,58 +3,72 @@
+};
 
 class TiledGalleryEdit extends Component {
 	constructor() {
```
This whole constructor can be removed.

sirreal commented on client/gutenberg/extensions/tiled-gallery/gallery-image.jsx:
```diff
@@ -0,0 +1,168 @@
+/**
+ * This component is originally from Gutenberg Gallery block:
+ * @link https://github.com/WordPress/gutenberg/blob/1cd604df7e9017e0dbe3ab64897ac3af35ca35c5/packages/block-library/src/gallery/gallery-image.js
+ */
```
We should try `import galleryImage from @wordpress/block-library/😬` at least to see what happens.

sirreal commented on client/gutenberg/extensions/tiled-gallery/view.js:
```diff
@@ -1,6 +1,134 @@
+	// Do nothing if page loads an editor
+	if ( isEditor ) {
+		return;
+	}
```
Do we need this?

sirreal commented on client/gutenberg/extensions/tiled-gallery/view.js:
```diff
@@ -1,6 +1,134 @@
+ * @returns {Array} List of gallery nodes on the page
+ */
+const getGalleries = () => {
+	return document ? [ ...document.querySelectorAll( '.wp-block-jetpack-tiled-gallery' ) ] : [];
```
Is this providing safety? Should likely be `'object' === typeof document`. If we keep this, lets use `isBrowser` defined above.

I question the utility of this method, we can likely inline it below and use `for...of`

sirreal commented on client/gutenberg/extensions/tiled-gallery/view.js:
```diff
@@ -1,6 +1,134 @@
+const isBrowser = typeof window !== 'undefined' && typeof document !== 'undefined';
+
+// @TODO this isn't working; need a reliable canonical way to determine editor vs view?
+const isEditor = isBrowser && window.hasOwnProperty( 'wp' ) && window.wp.hasOwnProperty( 'editor' );
```
This isn't enqueued automatically in the editor by Jetpack anymore, is it? We should be able to drop this now.

sirreal commented on client/gutenberg/extensions/tiled-gallery/view.js:
```diff
@@ -1,6 +1,134 @@
+ */
 import './view.scss';
+
+const isBrowser = typeof window !== 'undefined' && typeof document !== 'undefined';
```
We might match `=== 'object'`, but this is probably fine.

sirreal commented on client/gutenberg/extensions/tiled-gallery/view.js:
```diff
@@ -1,6 +1,134 @@
+	} );
+
+	// eslint-disable-next-line
+	// console.log('|', galleryLayout, options);
```
cleanup

sirreal commented on client/gutenberg/extensions/presets/jetpack/index-beta.json:
```diff
@@ -1,4 +1,5 @@
 [
   "related-posts",
+  "tiled-gallery",
```
Let's drop this for now and add it when we're in a better state.

simison replied:
Done https://github.com/Automattic/wp-calypso/pull/27458/commits/e3f65a73d1e86bc11c9683a205243477facacd1c

sirreal commented:
Was the shrinkwrap rebuilt recently? I noticed there have been several rebases and force pushes.

simison commented:
>Was the shrinkwrap rebuilt recently? I noticed there have been several rebases and force pushes.

Good call. I just rebased locally to confirm that shrinkwrap file is up to date.

sirreal commented:
Tests are all green now even though GH doesn't reflect that. Reporting mechanism hasn't been working well recently.

lezama commented:
🚢 IT 🎉

simison commented on client/gutenberg/extensions/tiled-gallery/view.js:
```diff
@@ -1,6 +1,117 @@
+ * Throttled resize event
+ * Compares gallery width to its previous width to decide if to resize it.
+ */
+const throttleOnResize = throttle( entries => {
```
I'm wondering if we would get better performance by using `debounce` and how that would look visually when resizing the window.

Differences: https://css-tricks.com/debouncing-throttling-explained-examples/

gsallman commented:
> > Captions - Should we include the animation in the block? I'd vote to drop that and do what ever core gallery does.
> 
> I agree with you. Let's drop those in favor of the way core gallery handles captions.

I am but a humble user putting in my opinion (with zero experience on Github, so apologies for any faux pas). If you are talking about how captions are rendered on the delivered page, then no, please do not do what the core gallery does. But if you are only talking about within the block editor, that's fine. 

The way the Core gallery handles the display of captions on the rendered page is unusable. (I don't know how it got past testing). I've tried it on several sites, and it is a responsive nightmare. In many cases the captions completely cover the images. Generally, I want visitors to be able to see all of the image, not have large portions of it obscured by text. But, I do want the caption text to be available for the curious (and search engines). On the other hand, the way the captions are currently handled in Jetpack with an animation is perfect, and shouldn't be changed. (Sorry for butting in here).

simison commented:
Thanks for the feedback, @gsallman. It's helpful.

Responsiveness of the Tiled gallery block is definitely on top of our priorities.
